docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+
+The `data` field contains the type-specific event datum, and `events` contain
+the ABCI events, if any were defined.
+
+> **Discussion point**: Based on [issue #7273][i7273], we may want to eliminate
+> this field from the response, since it duplicates data already stored
+> elsewhere in the event data.
+>
+> Alternatively, if clients need these data separately, we could require the
+> server to _move_ the events from the data message to this field, rather than
+> merely copying them.
+
+The semantics of the reply are as follows:
+
+- If `items` is non-empty:
+
+    - Items are ordered from newest to oldest.
+
+    - If `more` is true, there is at least one additional, older item in the
+      event log that was not returned (in excess of `max_results`).
+
+      In this case the client can fetch the next page by setting `before_item`
+      in a new request, to the cursor of the oldest item fetched (i.e., the
+      last one in `items`).
+
+    - Otherwise (if `more` is false), all the matching results have been
+      reported (pagination is complete).
+
+    - The first element of `items` identifies the newest item considered.
+      Subsequent poll requests can set `after_item` to this cursor to skip
+      items that were already retrieved.
+
+- If `items` is empty:
+
+    - If the `before_item` was set in the request, there are no further
+      eligible items for this query in the log (pagination is complete).
+
+      This is just a safety case; the client can detect this without issuing
+      another call by consulting the `more` field of the previous reply.
+
+    - If the `before_item` was empty in the request, no eligible items were
+      available before the `wait_time` expired. The client may poll again to
+      wait for more event items.
+
+A client can store cursor values to detect data loss and to recover from
+crashes and connectivity issues:
+
+- After a crash, the client requests events after the newest cursor it has
+  seen.  If the reply indicates that cursor is no longer in range, the client
+  may (conservatively) conclude some event data may have been lost.
+
+- On the other hand, if it _is_ in range, the client can then page back through
+  the results that it missed, and then resume polling. As long as its recovery
+  cursor does not age out before it finishes, the client can be sure it has all
+  the relevant results.
+
+### Other Notes
+
+- While the new API requires explicit polling by the client, it makes better
+  use of the node's existing HTTP infrastructure (e.g., connection pools).
+  Moreover, the direct implementation is easier to use from standard tools and
+  client libraries for HTTP and JSON-RPC.
+
+  Explicit polling does shift the burden of timeliness to the client.  That is
+  arguably preferable, however, given that the RPC service is ancillary to the
+  node's primary goal, viz., consensus. The details of polling can be easily
+  hidden from client applications with simple libraries.
```

Comment:
We already do this for all the cases where we get a http endpoint instead of a ws endpoint by th user.